File: kitchenos_nodejs/run_all.js
const { spawn } = require("child_process");
const path = require("path");

const SERVICES = [
  { name: "Metrics", script: "metrics_service.js" },
  { name: "Recipes", script: "recipes_service.js" },
  { name: "Pantry", script: "pantry_service.js" },
  { name: "Planner", script: "planner_service.js" },
  { name: "Shopping", script: "shopping_service.js" },
  { name: "Nutrition", script: "nutrition_service.js" },
  { name: "FrontendServer", script: "frontend_server.js" }
];

const children = [];
let isShuttingDown = false;

function streamLogs(name, stream) {
  stream.on("data", (data) => {
    const lines = data.toString("utf8").split("\n");
    lines.forEach((line) => {
      if (line.trim()) {
        console.log(`[${name}] ${line.trim()}`);
      }
    });
  });
}

function shutdown() {
  if (isShuttingDown) return;
  isShuttingDown = true;
  console.log("\nStopping all Node.js microservices...");
  
  children.forEach(({ name, proc }) => {
    if (proc.exitCode === null) {
      console.log(`  Killing ${name} (PID: ${proc.pid})...`);
      proc.kill("SIGTERM");
    }
  });
  console.log("All KitchenOS services stopped.\n");
  process.exit(0);
}

function main() {
  console.log("====================================================");
  console.log("     KitchenOS Node.js Platform Launcher            ");
  console.log("====================================================");
  console.log("Offline Mode: Persistent File-Based DB Client Active\n");
  console.log("Starting services...");

  SERVICES.forEach(({ name, script }) => {
    console.log(`  -> Starting ${name} (${script})...`);
    
    const scriptPath = path.join(__dirname, script);
    const proc = spawn("node", [scriptPath], {
      env: { ...process.env, PYTHONIOENCODING: "utf-8" },
      stdio: ["ignore", "pipe", "pipe"]
    });

    children.push({ name, proc });

    streamLogs(name, proc.stdout);
    streamLogs(name + " [Err]", proc.stderr);

    proc.on("exit", (code) => {
      if (!isShuttingDown) {
        console.log(`\n[Warning] ${name} service terminated with exit code ${code}!`);
        shutdown();
      }
    });
  });

  console.log("\nAll microservices and local frontend server initialized successfully.");
  console.log("Central Access Endpoint: http://localhost:3000");
  console.log("Press Ctrl+C to terminate all services.\n");

  // Keep process alive and handle termination signals
  process.on("SIGINT", shutdown);
  process.on("SIGTERM", shutdown);
  
  setInterval(() => {}, 1000); // keep event loop busy
}

main();

File: autohub_python/run_all.py
import subprocess
import sys
import threading
import time
import os

SERVICES = [
    ("Metrics", "metrics_service.py"),
    ("Auth", "auth_service.py"),
    ("Inventory", "inventory_service.py"),
    ("Booking", "booking_service.py"),
    ("Maintenance", "maintenance_service.py"),
    ("Payment", "payment_service.py"),
    ("Valuation", "valuation_service.py"),
    ("Fuel", "fuel_service.py"),
    ("Insurance", "insurance_service.py"),
    ("FrontendServer", "frontend_server.py")
]

processes = []

def stream_logs(name, proc):
    try:
        while True:
            line = proc.stdout.readline()
            if not line:
                break
            print(f"[{name}] {line.decode('utf-8', errors='ignore').strip()}")
            sys.stdout.flush()
    except Exception as e:
        print(f"[{name}] Log stream error: {e}")

def main():
    print("====================================================")
    print("      AutoHub Microservices Platform Launcher       ")
    print("====================================================")
    
    mongo_uri = os.environ.get("MONGO_URI")
    if mongo_uri:
        print("Production Mode: Connecting to AWS / Remote Database")
    else:
        print("Offline Mode: Persistent File-Based DB Client Active")
        
    print("\nStarting services...")

    # Start each service in sequence
    for name, script in SERVICES:
        print(f"  -> Starting {name} ({script})...")
        try:
            proc = subprocess.Popen(
                [sys.executable, "-u", script],
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                bufsize=1
            )
            processes.append((name, proc))
            
            t = threading.Thread(target=stream_logs, args=(name, proc), daemon=True)
            t.start()
            
            time.sleep(0.5)
        except Exception as e:
            print(f"Failed to start {name}: {e}")
            shutdown()
            sys.exit(1)

    print("\nAll microservices and local frontend server initialized successfully.")
    print("Central Access Endpoint: http://localhost:5000")
    print("Press Ctrl+C to terminate all services.\n")

    # Keep main thread alive
    try:
        while True:
            time.sleep(1)
            for name, proc in processes:
                poll = proc.poll()
                if poll is not None:
                    print(f"\n[Warning] {name} service terminated with exit code {poll}!")
                    shutdown()
                    sys.exit(1)
    except KeyboardInterrupt:
        print("\nCtrl+C detected. Initiating graceful shutdown...")
        shutdown()

def shutdown():
    print("\nStopping all microservices...")
    for name, proc in processes:
        if proc.poll() is None:
            print(f"  Killing {name} (PID: {proc.pid})...")
            try:
                proc.terminate()
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                proc.kill()
            except Exception as e:
                print(f"Error stopping {name}: {e}")
    print("All services stopped. AutoHub offline.\n")

if __name__ == "__main__":
    main()
